src/controllers/settings.js

```diff
@@ -2,16 +2,199 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+import AppConstants from '../app-constants.js'
+
+import {
+  getUserEmails,
+  resetUnverifiedEmailAddress,
+  addSubscriberUnverifiedEmailHash,
+  removeOneSecondaryEmail,
+  getEmailById,
+  verifyEmailHash
+} from '../db/tables/email_addresses.js'
+
+import { setAllEmailsToPrimary } from '../db/tables/subscribers.js'
+
+import { fluentError, getMessage } from '../utils/fluent.js'
+import { sendEmail, getVerificationUrl, getUnsubscribeUrl } from '../utils/email.js'
+
+import { getBreachesForEmail } from '../utils/hibp.js'
+import { generateToken } from '../utils/csrf.js'
+
 import { mainLayout } from '../views/main.js'
 import { settings } from '../views/partials/settings.js'
+import { getTemplate } from '../emails/templates/email-2022.js'
+import { verifyPartial } from '../emails/partials/email-verify.js'
+
+async function settingsPage (req, res) {
+  const emails = await getUserEmails(req.session.user.id)
+  // Add primary subscriber email to the list
+  emails.push({
+    email: req.session.user.primary_email,
+    sha1: req.session.user.primary_sha1,
+    primary: true,
+    verified: true
+  })
+
+  const breachCounts = new Map()
+
+  const allBreaches = req.app.locals.breaches
+  for (const email of emails) {
+    const breaches = await getBreachesForEmail(
+      email.sha1,
+      allBreaches,
+      true,
+      false
+    )
+    breachCounts.set(email.email, breaches?.length || 0)
+  }
 
-function settingsPage (req, res) {
   const data = {
     fxaProfile: req.user.fxa_profile_json,
-    partial: settings
+    partial: settings,
+    emails,
+    breachCounts,
+    limit: AppConstants.MAX_NUM_ADDRESSES,
+    csrfToken: generateToken(res)
   }
 
   res.send(mainLayout(data))
 }
 
-export { settingsPage }
+async function addEmail (req, res) {
+  const sessionUser = await req.user
+  const email = req.body.email
+  // Use the same regex as HTML5 email input type
+  // https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/email#basic_validation
+  const emailRegex = /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/
+
+  if (!email || !email.match(emailRegex)) {
+    throw fluentError('user-add-invalid-email')
+  }
+
+  if (sessionUser.email_addresses.length >= AppConstants.MAX_NUM_ADDRESSES) {
+    throw fluentError('user-add-too-many-emails')
+  }
+
+  _checkForDuplicateEmail(sessionUser, email)
+
+  const unverifiedSubscriber = await addSubscriberUnverifiedEmailHash(
+    req.session.user,
+    email
+  )
+
+  await _sendVerificationEmail(unverifiedSubscriber.id)
+
+  return res.json({
+    success: true,
+    status: 200,
+    message: 'Sent the verification email'
+  })
+}
+
+function _checkForDuplicateEmail (sessionUser, email) {
+  email = email.toLowerCase()
```

Comment:
Suggestion (non-blocking): Let’s try to avoid mutating passed args.

```suggestion
  const emailLowerCase = email.toLowerCase()
```